src/types/acls.rs

```diff
@@ -0,0 +1,398 @@
+    InvalidRight,
+}
+
+impl Display for AclRightError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match *self {
+            AclRightError::InvalidRight => {
+                write!(f, "Rights may only be lowercase alpha numeric characters")
+            }
+        }
+    }
+}
+
+impl std::error::Error for AclRightError {}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The ACL response from the getacl IMAP command
+#[self_referencing]
+pub struct Acl {
+    data: Vec<u8>,
+    #[borrows(data)]
+    #[covariant]
+    pub(crate) acl: InnerAcl<'this>,
+}
+
+impl Acl {
+    /// Parse the given input into a [`ACL`] response.
+    pub fn parse(
+        owned: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        AclTryBuilder {
+            data: owned,
+            acl_builder: |input| {
+                let mut lines: &[u8] = input;
+
+                // There should only be ONE single ACL response
+                while !lines.is_empty() {
+                    match imap_proto::parser::parse_response(lines) {
+                        Ok((_rest, Response::Acl(a))) => {
+                            // lines = rest;
+                            return Ok(InnerAcl {
+                                mailbox: a.mailbox,
+                                acls: a
+                                    .acls
+                                    .into_iter()
+                                    .map(|e| AclEntry {
+                                        identifier: e.identifier,
+                                        rights: e.rights.into(),
+                                    })
+                                    .collect(),
+                            });
+                        }
+                        Ok((rest, data)) => {
+                            lines = rest;
+                            if let Some(resp) = try_handle_unilateral(data, unsolicited) {
+                                return Err(resp.into());
+                            }
+                        }
+                        _ => {
+                            return Err(Error::Parse(ParseError::Invalid(lines.to_vec())));
+                        }
+                    }
+                }
+
+                Err(Error::Parse(ParseError::Invalid(lines.to_vec())))
+            },
+        }
+        .try_build()
+    }
+
+    /// Return the mailbox the ACL entries belong to
+    pub fn mailbox(&self) -> &str {
+        &*self.borrow_acl().mailbox
+    }
+
+    /// Returns a list of identifier/rights pairs for the mailbox
+    pub fn acls(&self) -> &[AclEntry<'_>] {
+        &*self.borrow_acl().acls
+    }
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct InnerAcl<'a> {
+    /// The mailbox the ACL Entries belong to
+    pub(crate) mailbox: Cow<'a, str>,
```

Comment:
I'm a little confused about why this and other fields are `Cow` and not just `&'a str` when it doesn't appear they're ever `Cow::Owned`, and will never be settable directly by the user?

Reply:
If I set them as `&'a str`, then the ouroboros builders fail to compile as the Cow is how they become referenced against the internal data struct. So there technically is an "owned" but it's just a transfer from the imap-proto type to my type (see line 127 in `Acl::parse`)

Reply:
Huh, that's very odd. I wonder if we could just match on the `Cow` to specifically grab the `'a` reference out of the `Cow::Borrowed` if we know it's always that variant. Or, phrased differently, I don't quite see how it'd ever end up being a `Cow::Owned` in that code, since the parser should already only be yielding things that borrow from `input` (I think). Unless it _sometimes_ does decoding I suppose, which could be the case.